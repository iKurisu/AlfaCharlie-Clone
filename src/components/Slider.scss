.slider-mask {
  width: 100%;
  height: 100%;
  background: #fff;
  position: absolute;
  z-index: 10; 
}

.slider-wrapper,
.slide {
  width: 100%;
  height: 100%;
}

.slider-wrapper {
  display: flex;
}

.slide {
  position: relative;
  overflow: hidden;
  flex-shrink: 0;
}

.slide-img {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-size: cover;
  background-position: center;
}
