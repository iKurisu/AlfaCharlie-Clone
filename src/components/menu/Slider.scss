.menu-images {
  width: 50%;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
  z-index: 10;
}

.menu-slider {
  width: 27.2vw;
  height: 70vh;
  clip-path: inset(1px 0 0 1px);
  position: relative;
}

.slider-mask {
  width: 100%;
  height: 100%;
  background: #fff;
  position: absolute;
  z-index: 10; 
}

.slider-wrapper,
.slide {
  width: 100%;
  height: 100%;
}

.slider-wrapper {
  display: flex;
  // overflow: hidden;
  // transition: transform 1s;
}

.slide {
  position: relative;
  overflow: hidden;
  flex-shrink: 0;
  width: 399px;
}

.slide-img {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-size: cover;
  background-position: center;
  // transition: transform 1s;
}

@media (max-width: 1024px) {
  .menu-images {
    display: none;
  }
}

@media (max-width: 1600px) {
  .slide {
    width: 100%;
  }
}
