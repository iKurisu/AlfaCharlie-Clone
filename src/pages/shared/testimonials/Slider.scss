.testimonials-content .testimonials-slider {
  position: relative;
  height: 375px;
}

.testimonials-content .testimonials-slide {
  position: absolute;
  top: 0;
  transition: opacity 1.3s ease;
}

.testimonials-content .testimonials-slide:not(.active) {
  opacity: 0;
  visibility: hidden;
}

.testimonial {
  width: 88%;
  margin-bottom: 30px;
}

.testimonial p {
  font-family: "benton-modern-display";
  font-size: 25px;
  letter-spacing: 0.08em;
  line-height: 1.327;
}
