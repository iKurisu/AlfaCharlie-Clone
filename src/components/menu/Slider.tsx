import React, { useState, useRef, useEffect } from "react";
import useTransition from "hooks/useTransition";
import useDidUpdateEffect from "hooks/useDidUpdateEffect";
import { vwToPx } from "utils/responsive";
import "./Slider.scss";

const urlSite = "https://alfacharlie.b-cdn.net/wp-content/uploads/";
const imageUrls = [
  "2019/05/Alfa-Charlie-Creative-Agency-home-e1558112927714.jpg",
  "2019/05/Alfa-Charlie-Creative-Agency-about-5.jpg",
  "2019/04/Alfa-Charlie-Creative-Agency-work.jpg",
  "2019/05/Alfa-Charlie-Creative-Agency-home-2.jpg",
  "2019/05/Alfa-Charlie-Creative-Agency-contact-2.jpg"
];

interface Props {
  isOpen: boolean;
  hoveringElementId: number;
  previousElementId: number;
}

const Slider = ({
  isOpen,
  hoveringElementId,
  previousElementId
}: Props): JSX.Element => {
  const [wrapperTranslate, setWrapperTranslate] = useState(vwToPx(27.2));
  const [imageTranslate, setImageTranslate] = useState(vwToPx(23.15));

  const updateTranslateValues = (): void => {
    setWrapperTranslate(vwToPx(27.2));
    setImageTranslate(vwToPx(23.15));
  };

  useEffect((): (() => void) => {
    window.addEventListener("resize", updateTranslateValues);

    return (): void =>
      window.removeEventListener("resize", updateTranslateValues);
  }, [imageTranslate]);

  const wrapper = useRef(null);
  const mask = useRef(null);

  const fadeIn = useTransition(wrapper, {
    from: { opacity: 0 },
    to: { opacity: 1 },
    config: {
      delay: 800,
      duration: 400
    }
  });

  const fadeOut = useTransition(wrapper, {
    from: { opacity: 1 },
    to: { opacity: 0 },
    config: {
      delay: 650,
      duration: 0
    }
  });

  const revealSlider = useTransition(mask, {
    from: { transform: "translateX(0)" },
    to: { transform: "translateX(100%)" },
    config: {
      duration: 850,
      delay: 800,
      timing: [0.17, 0.5, 0.48, 1]
    }
  });

  const hideSlider = useTransition(mask, {
    from: { transform: "translateX(-100%)" },
    to: { transform: "translateX(0)" },
    config: {
      duration: 1000,
      timing: [0.17, 0.5, 0.48, 1]
    }
  });

  useDidUpdateEffect((): void => {
    if (isOpen) {
      fadeIn();
      revealSlider();
    } else {
      hideSlider();
      fadeOut();
    }
  }, [isOpen]);

  const transitionLength =
    Math.abs(previousElementId - hoveringElementId) === 1 ? "1s" : "2s";

  return (
    <div className="menu-images -flex" ref={wrapper} style={{ opacity: 0 }}>
      <div className="menu-slider">
        <div className="slider-mask" ref={mask} />
        <div
          className="slider-wrapper"
          style={{
            transform: `translateX(${hoveringElementId * -wrapperTranslate}px)`,
            transition: `transform ${transitionLength}`
          }}
        >
          {imageUrls.map(
            (url, id): JSX.Element => (
              <div className="slide" key={id}>
                <div
                  className="slide-img"
                  style={{
                    backgroundImage: `url(${urlSite}${url})`,
                    transform: `translateX(${1 -
                      wrapperTranslate * (id - hoveringElementId)}px)`,
                    transition: `transform ${transitionLength}`
                  }}
                />
              </div>
            )
          )}
        </div>
      </div>
    </div>
  );
};

export default Slider;
