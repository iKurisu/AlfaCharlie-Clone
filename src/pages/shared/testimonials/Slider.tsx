import React from "react";
import { connect } from "react-redux";
import { AppState } from "store";
import Link from "../Link";
import { classList } from "utils/class";
import "./Slider.scss";

export interface Author {
  name: string;
  position: string;
}

interface Testimonial {
  paragraphs: string[];
  author: Author;
}

const testimonials: Testimonial[] = [
  {
    paragraphs: [
      "Since we began working with Alfa Charlie, they have elevated our brand " +
        "through creative web and print design projects. Their ability to listen " +
        "to our marketing goals and understand our company has brought on " +
        "extremely satisfying results making us feel like Alfa Charlie is part " +
        "of our team. Not only fun to work with, Alfa Charlie is responsive, " +
        "organized and produces beautiful work."
    ],
    author: {
      name: "Edward Piegza",
      position: "Founder, Classic Journeys"
    }
  },
  {
    paragraphs: [
      "We get so many compliments on our website! Alfa Charlie did an amazing " +
        "job, they were able to depict our brand and purpose through a fun, " +
        "clean and professional website. If you’re looking for a service that " +
        "will take the extra time to get to know your company personally and " +
        "create a website centered around your brand, goals, and purpose, then " +
        "Alfa Charlie is for you."
    ],
    author: {
      name: "Teal Cooper",
      position: "Founder, Vendibean"
    }
  },
  {
    paragraphs: [
      "Working with Alfa Charlie was a dream! They spent time really getting a " +
        "feel for where I wanted to take my business in the coming years and " +
        "what type of client I wanted to attract.",
      "And the result was stunning! I look at my website every day and think " +
        "about how it reflects my style, ideal client, and business goals perfectly!"
    ],
    author: {
      name: "Kristina Doherty",
      position: "Founder, Kristina Kay Photography"
    }
  },
  {
    paragraphs: [
      "A head and shoulders above others, Alfa Charlie is a leading design " +
        "agency. As the owner of a growing communications firm, I need to " +
        "partner with teams that are quick, reliable and creative. On every " +
        "project, Alfa Charlie has delivered an exceptional product and " +
        "service for our clients that we’ve all been very proud of."
    ],
    author: {
      name: "Charles Chamberlayne",
      position: "Principal, Chamberlaynepr"
    }
  },
  {
    paragraphs: [
      "The creative process with Kim and Reva was both interactive and fun, " +
        "and the quality of their work is second to none. I couldn’t be happier " +
        "with the outcome. Having a website and brand that stands out from my " +
        "competition and represents my personality has really given me great " +
        "confidence in a time that would otherwise have been daunting. I highly " +
        "recommend their services."
    ],
    author: {
      name: "Jono Green",
      position: "Founder, GRX Baseball"
    }
  }
];

interface MappedState {
  activeSlide: number;
}

export const Slider = ({ activeSlide }: MappedState): JSX.Element => (
  <div className="testimonials-slider">
    {testimonials.map(({ paragraphs, author }, id) => (
      <div
        className={classList({
          ["testimonials-slide"]: true,
          active: id === activeSlide
        })}
        key={id}
      >
        <div className="testimonial">
          {paragraphs.map((paragraph, id) => (
            <p key={id}>{paragraph}</p>
          ))}
        </div>
        <div className="testimonial-author">
          <Link content={author} />
        </div>
      </div>
    ))}
  </div>
);

const mapState = ({ testimonials }: AppState): MappedState => ({
  activeSlide: testimonials.currentSlideID
});

export default connect(mapState)(Slider);
