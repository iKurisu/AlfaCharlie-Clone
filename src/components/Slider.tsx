import React, { useRef, MouseEventHandler, RefObject } from "react";
import { Dispatch } from "redux";
import { connect } from "react-redux";
import { AppState } from "store";
import { menuActions } from "modules/menu";
import { heroActions } from "modules/hero";
import Mask from "./slider/Mask";
import Slide from "./slider/Slide";
import useResponsiveWidth from "hooks/useResponsiveWidth";
import useDrag, { Handler } from "hooks/useDrag";
import { getDistance, getDuration } from "utils/slider";
import { setTransform, setTransition } from "utils/refs";
import "./Slider.scss";

interface MappedState {
  isOpen: boolean;
  currentSlideID: number;
  previousSlideID: number;
}

interface MappedActions {
  swipeSlide: (elementID: number, duration: number) => MouseEventHandler;
}

interface SliderOptions {
  fadeDirection: "left" | "right";
  vw: {
    image: number;
    wrapper: number;
  };
  delay?: number;
  maxLength?: number;
}

interface OwnProps {
  imageUrls: string[];
  options: SliderOptions;
}

type Props = MappedState & MappedActions & OwnProps;

interface Properties {
  transform?: (id: number) => string;
  transition?: string;
}

enum MouseDirection {
  LEFT = 1,
  RIGHT = -1
}

export const Slider = ({
  imageUrls,
  options: { fadeDirection, delay = 0, vw, maxLength = 3000 },
  isOpen,
  currentSlideID,
  previousSlideID,
  swipeSlide
}: Props): JSX.Element => {
  const wrapperWidth = useResponsiveWidth(vw.wrapper);
  const imageWidth = useResponsiveWidth(vw.image);

  const wrapper = useRef(null);
  const images = imageUrls.map(() => useRef(null));

  const wrapperDistance = currentSlideID * -wrapperWidth;
  const getSlideDistance = (id: number): number =>
    1 - imageWidth * (id - currentSlideID);

  const setWrapperTransform = setTransform(wrapper);
  const setWrapperTransition = setTransition(wrapper);

  const updateProperties = ({ transform, transition }: Properties) => (
    image: RefObject<HTMLDivElement>,
    id: number
  ) => {
    if (transform) setTransform(image)(transform(id));
    if (transition) setTransition(image)(transition);
  };

  const onTransition = useRef(false);

  const toggleTransition = (): void => {
    onTransition.current = true;
    setTimeout(() => (onTransition.current = false), 1000);
  };

  const isOnTransition = (): boolean => onTransition.current;

  const isFirstSlide = (mouseX: number, clickPosition: number): boolean =>
    currentSlideID === 0 && mouseX > clickPosition;
  const isLastSlide = (mouseX: number, clickPosition: number): boolean =>
    currentSlideID === 4 && mouseX < clickPosition;

  const isAtExtreme: typeof isFirstSlide = (...args) =>
    isFirstSlide(...args) || isLastSlide(...args);

  const canSwipe: typeof isFirstSlide = (...args) =>
    !isAtExtreme(...args) && !isOnTransition();

  const onDrag: Handler = (event, clickPosition) => {
    if (!canSwipe(event.clientX, clickPosition)) return;

    setWrapperTransform(
      `translateX(${wrapperDistance + event.clientX - clickPosition}px)`
    );

    images.forEach(
      updateProperties({
        transform: (id: number): string => {
          return `translateX(${1 -
            imageWidth * (id - (clickPosition - event.clientX) / wrapperWidth) +
            imageWidth * currentSlideID}px)`;
        }
      })
    );
  };

  const onDrop: Handler = (event, clickPosition) => {
    if (!canSwipe(event.clientX, clickPosition)) return;

    if (getDistance(event.clientX, clickPosition) < wrapperWidth * 0.2) {
      toggleTransition();

      setWrapperTransform(`translateX(${wrapperDistance}px)`);
      setWrapperTransition("transform 1s");

      images.forEach(
        updateProperties({
          transform: (id: number): string =>
            `translateX(${getSlideDistance(id)}px)`,
          transition: "transform 1s"
        })
      );

      setTimeout(() => {
        setWrapperTransition("transform 0s");

        images.forEach(updateProperties({ transition: "transform 0s" }));
      }, 1000);
    } else {
      const direction =
        event.clientX < clickPosition
          ? MouseDirection.LEFT
          : MouseDirection.RIGHT;

      const duration = getDuration({
        from: currentSlideID,
        to: currentSlideID + direction,
        max: maxLength
      });

      toggleTransition();
      swipeSlide(currentSlideID + direction, duration)(event);
    }
  };

  const dragProps = useDrag({ onDrag, onDrop });

  const transitionDuration = getDuration({
    from: previousSlideID,
    to: currentSlideID,
    max: maxLength
  });

  return (
    <React.Fragment>
      <Mask isOpen={isOpen} options={{ fadeDirection, delay }} />
      <div className="slider-swiper" {...dragProps}>
        <div
          className="slider-wrapper"
          ref={wrapper}
          style={{
            transform: `translateX(${wrapperDistance.toFixed(3)}px)`,
            transition: `transform ${transitionDuration}ms`
          }}
        >
          {imageUrls.map(
            (url, id): JSX.Element => (
              <Slide
                url={url}
                distance={getSlideDistance(id)}
                duration={transitionDuration}
                imageRef={images[id]}
                key={id}
              />
            )
          )}
        </div>
      </div>
    </React.Fragment>
  );
};

const mapMenuState = ({ menu }: AppState): MappedState => ({
  isOpen: menu.toggled,
  currentSlideID: menu.hoveringElementID,
  previousSlideID: menu.previousElementID
});

const mapMenuDispatch = (dispatch: Dispatch): MappedActions => ({
  swipeSlide: (
    elementID: number,
    duration: number
  ): React.MouseEventHandler => (): void => {
    dispatch(menuActions.setHoveringElement(elementID));
    setTimeout(() => dispatch(menuActions.updatePreviousElement()), duration);
  }
});

export const MenuSlider = connect(
  mapMenuState,
  mapMenuDispatch
)(Slider);

const mapHeroState = ({ hero, intro }: AppState): MappedState => ({
  isOpen: !intro.toggled,
  currentSlideID: hero.currentSlideID,
  previousSlideID: hero.previousSlideID
});

const mapHeroDispatch = (dispatch: Dispatch): MappedActions => ({
  swipeSlide: (
    elementID: number,
    duration: number
  ): React.MouseEventHandler => (): void => {
    dispatch(heroActions.setSlide(elementID));
    setTimeout(() => dispatch(heroActions.updatePreviousSlide()), duration);
  }
});

export const HeroSlider = connect(
  mapHeroState,
  mapHeroDispatch
)(Slider);
