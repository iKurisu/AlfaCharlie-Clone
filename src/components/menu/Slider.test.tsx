import React from "react";
import renderer from "react-test-renderer";
import Slider from "./Slider";

it("renders correctly", (): void => {
  const component = renderer.create(<Slider />);
  expect(component).toMatchSnapshot();
});
