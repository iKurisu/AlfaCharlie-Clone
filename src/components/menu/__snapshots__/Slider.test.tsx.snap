// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="menu-images -flex"
  style={
    Object {
      "opacity": 0,
    }
  }
>
  <div
    className="menu-slider"
  >
    <div
      className="slider-mask"
    />
    <div
      className="slider-wrapper"
      style={
        Object {
          "transform": "translateX(0px)",
          "transition": "transform 2s",
        }
      }
    >
      <div
        className="slide"
      >
        <div
          className="slide-img"
          style={
            Object {
              "backgroundImage": "url(https://alfacharlie.b-cdn.net/wp-content/uploads/2019/05/Alfa-Charlie-Creative-Agency-home-e1558112927714.jpg)",
              "transform": "translateX(1px)",
              "transition": "transform 2s",
            }
          }
        />
      </div>
      <div
        className="slide"
      >
        <div
          className="slide-img"
          style={
            Object {
              "backgroundImage": "url(https://alfacharlie.b-cdn.net/wp-content/uploads/2019/05/Alfa-Charlie-Creative-Agency-about-5.jpg)",
              "transform": "translateX(1px)",
              "transition": "transform 2s",
            }
          }
        />
      </div>
      <div
        className="slide"
      >
        <div
          className="slide-img"
          style={
            Object {
              "backgroundImage": "url(https://alfacharlie.b-cdn.net/wp-content/uploads/2019/04/Alfa-Charlie-Creative-Agency-work.jpg)",
              "transform": "translateX(1px)",
              "transition": "transform 2s",
            }
          }
        />
      </div>
      <div
        className="slide"
      >
        <div
          className="slide-img"
          style={
            Object {
              "backgroundImage": "url(https://alfacharlie.b-cdn.net/wp-content/uploads/2019/05/Alfa-Charlie-Creative-Agency-home-2.jpg)",
              "transform": "translateX(1px)",
              "transition": "transform 2s",
            }
          }
        />
      </div>
      <div
        className="slide"
      >
        <div
          className="slide-img"
          style={
            Object {
              "backgroundImage": "url(https://alfacharlie.b-cdn.net/wp-content/uploads/2019/05/Alfa-Charlie-Creative-Agency-contact-2.jpg)",
              "transform": "translateX(1px)",
              "transition": "transform 2s",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
